lib/ansible/plugins/callback/grafana_annotations.py

```diff
@@ -0,0 +1,234 @@
+      http_agent:
+        description: The HTTP 'User-agent' value to set in HTTP requets.
+        env:
+          - name: HTTP_AGENT
+        default: 'Ansible (grafana_annotations callback)'
+      api_key:
+        description: Grafana API key, allowing to authenticate when posting on the HTTP API.
+                     If not provided, grafana_login and grafana_password will
+                     be required.
+        env:
+          - name: GRAFANA_API_KEY
+      grafana_user:
+        description: Grafana user used for authentication. Ignored if api_key is provided.
+        env:
+          - name: GRAFANA_USER
+      grafana_password:
+        description: Grafana password used for authentication. Ignored if api_key is provided.
+        env:
+          - name: GRAFANA_PASSWORD
+      grafana_dashboard_id:
+        description: The grafana dashboard id where the annotation shall be created.
+        env:
+          - name: GRAFANA_DASHBOARD_ID
+      grafana_panel_id:
+        description: The grafana panel id where the annotation shall be created.
+        env:
+          - name: GRAFANA_PANEL_ID
+'''
+
+
+PLAYBOOK_START_TXT = """\
+Started playbook {playbook}
+
+From '{hostname}'
+By user '{username}'
+"""
+
+PLAYBOOK_ERROR_TXT = """\
+Playbook {playbook} Failure !
+
+From '{hostname}'
+By user '{username}'
+
+'{task}' failed on {host}
+
+debug: {result}
+"""
+
+PLAYBOOK_STATS_TXT = """\
+Playbook {playbook}
+Duration: {duration}
+Status: {status}
+
+From '{hostname}'
+By user '{username}'
+
+Result:
+{summary}
+"""
+
+
+def to_millis(dt):
+    return int(dt.strftime('%s')) * 1000
+
+
+def str2bool(string):
+    if string in [True, 'True', 'true', '1', 'yes']:
+        return True
+    elif string in [False, 'False', 'false', '0', 'no']:
+        return True
+    else:
+        raise Exception("Unsupported value '%s' as boolean" % string)
+
+
+class CallbackModule(CallbackBase):
+    """
+    ansible grafana callback plugin
+    ansible.cfg:
+        callback_plugins   = <path_to_callback_plugins_folder>
+        callback_whitelist = grafana_annotations
+    and put the plugin in <path_to_callback_plugins_folder>
+    """
+
+    CALLBACK_VERSION = 1.0
+    CALLBACK_TYPE = 'aggregate'
+    CALLBACK_NAME = 'grafana_annotations'
+    CALLBACK_NEEDS_WHITELIST = True
+
+    def __init__(self):
+        super(CallbackModule, self).__init__()
+
+        api_key = os.getenv('GRAFANA_API_KEY', None)
+        self.grafana_url = os.getenv('GRAFANA_URL', 'http://127.0.0.1:3000/api/annotations')
+        self.grafana_validate_certs = str2bool(os.getenv('GRAFANA_VALIDATE_CERT', True))
+        self.http_agent = os.getenv('HTTP_AGENT', 'Ansible (grafana_annotations callback)')
+        self.grafana_user = os.getenv('GRAFANA_USER', None)
+        self.grafana_password = os.getenv('GRAFANA_PASSWORD', None)
+        self.dashboard_id = os.getenv('GRAFANA_DASHBOARD_ID', None)
+        self.panel_id = os.getenv('GRAFANA_PANEL_ID', None)
```

Comment:
These `os.getenv` calls should likely be updated to use `self.get_option` and potentially `self.set_options`.  See the `slack` callback for examples.

Reply:
Without this change, the `DOCUMENTATION` isn't really being used, and is likely to cause confusion with users who assume they can set these like any other ansible config value.

Reply:
Hi @sivel, thanks for the review.
After a quick read of the code, I'm not sure to fully understand the get_option a DOCUMENTATION relation. 

From the slack example below:
```
    options:
      webhook_url:
        required: True
        description: Slack Webhook URL
        env:
          - name: SLACK_WEBHOOK_URL
        ini:
          - section: callback_slack
            key: webhook_url
```
Is the DOCUMENTATION string parsed to retrieve the env variable name that can contain the value and/or the ansible.cfg section and parameter key ?

Is there a clear documentation explaining this that I would have missed ?

Reply:
Yes, DOCUMENTATION is parsed to define configuration values for your plugin.  Any values supplied via ansible.cfg or env var will be made available to your plugin through the `set_options`/`get_option` API.

The slack callback makes use of both `set_options` and `get_option`